llrt_core/src/modules/timers.rs

```diff
@@ -15,103 +17,143 @@ use rquickjs::{
     cb: Function<'js>,
-    delay: usize,
+    delay: u64,
     repeating: bool,
 ) -> Result<usize> {
-    let expires = get_current_time_millis() + delay;
+    let deadline = Instant::now() + Duration::from_millis(delay);
     let id = TIMER_ID.fetch_add(1, Ordering::Relaxed);
 
     let callback = Persistent::<Function>::save(ctx, cb);
 
-    let rt = unsafe { qjs::JS_GetRuntime(ctx.as_raw().as_ptr()) };
-    let mut rt_timers = RUNTIME_TIMERS.lock().unwrap();
-
-    let timeout_ref = TimeoutRef {
-        expires,
+    let timeout = Timeout {
+        deadline,
         callback: Some(callback),
-        ctx: ctx.as_raw(),
+        raw_ctx: ctx.as_raw(),
         id,
         repeating,
-        delay,
+        interval: delay,
     };
 
-    if let Some(entry) = rt_timers.iter_mut().find(|state| state.rt == rt) {
-        entry.timers.push(timeout_ref);
+    let rt_ptr = unsafe { qjs::JS_GetRuntime(ctx.as_raw().as_ptr()) };
+
+    let mut rt_timer = RT_TIMER_STATE.lock().unwrap();
+    let state = get_timer_state(&mut rt_timer, rt_ptr);
+    state.timers.push(timeout);
+    let mut abort_timer_rx = None;
+    let task_running = state.running;
+    if task_running {
+        if deadline < state.deadline {
+            state.deadline = deadline;
+            state.notify.notify_one();
+        }
     } else {
-        let mut entry = RuntimeTimerState::new(rt);
-        entry.timers.push(timeout_ref);
-        rt_timers.push(entry);
+        state.running = true;
+        abort_timer_rx = Some(state.notify.clone());
+    }
+    drop(rt_timer);
+
+    if !task_running {
+        create_spawn_loop(
+            rt_ptr,
+            ctx,
+            unsafe { abort_timer_rx.unwrap_unchecked() },
+            deadline,
+        )?;
     }
-    drop(rt_timers);
-    create_spawn_loop(ctx, rt)?;
 
     Ok(id)
 }
 
+fn get_timer_state<'a>(
+    state_ref: &'a mut MutexGuard<Vec<RuntimeTimerState>>,
+    rt: *mut qjs::JSRuntime,
+) -> &'a mut RuntimeTimerState {
+    let rt_timers = state_ref.iter_mut().find(|state| state.rt == rt);
+
+    (unsafe { rt_timers.unwrap_unchecked() }) as _
+}
+
 fn clear_timeout_interval(ctx: Ctx<'_>, id: usize) -> Result<()> {
     let rt = unsafe { qjs::JS_GetRuntime(ctx.as_raw().as_ptr()) };
-    let mut rt_timers = RUNTIME_TIMERS.lock().unwrap();
+    let mut rt_timers = RT_TIMER_STATE.lock().unwrap();
 
-    if let Some(state) = rt_timers.iter_mut().find(|t| t.rt == rt) {
-        if let Some(timeout) = state.timers.iter_mut().find(|t| t.id == id) {
-            if let Some(timeout) = timeout.callback.take() {
-                timeout.restore(&ctx)?; //prevent memory leaks
-            }
-            timeout.expires = 0;
-            timeout.repeating = false;
+    let state = get_timer_state(&mut rt_timers, rt);
+    if let Some(timeout) = state.timers.iter_mut().find(|t| t.id == id) {
+        if let Some(timeout) = timeout.callback.take() {
+            timeout.restore(&ctx)?; //prevent memory leaks
```

Comment:
A drop is sufficient, restore is just some fancy lifetime changes. The drop trait of Value is what decreases the ref count.